### DISCUSSION TITLE
What is `bind` and `transform` in the context of `Variable` and `Service`?

### DISCUSSION THREAD

**Discussion by matheushpmoreira** [Aylur/ags#290] [Q&A] [answered]
`Widget.bind` obviously just sets an event listener, so it's very straightforward to understand. They clearly are not the same methods as defined in `Widget`, since `Variable.bind` requires no arguments and `Service.bind` only needs the property name. However, both seem to return something that... resembles a pointer? I know it's not one, though. I don't know how a property expecting a string is able to take this "pointer-like object".

I read the entire documentation, so it might have gone over my head, but the pages for `Variable` and `Service` don't really explain what exactly these methods do, only how to place them in code with examples - which is fine and comprehensible, I'm just being curious. I was able to find their definitions on the source code, and it looks like `bind` is defined with some kind of chain of generics. I've barely learned JavaScript, much less TypeScript, so it's impossible for me to make out what it's doing.

**Comment by Aylur** [comment, marked as answer]
both of them return an instance of [Binding](https://github.com/Aylur/ags/blob/bbbf87a21705b7566c2361bc1b978517530d10c0/src/service.ts#L14) which is basically a struct that holds
- the `emitter` object that needs to be watched
- its property
- a `transformFn` which is not meant to be called, only set, is a function that the value will go through and be set on a widget property
- a callable transform method which returns a new instance of `Binding` with an updated `transfromFn`

it is possible to chain `transform`s since they will be composed, I think they are called composite functions in math

this `Binding` struct is only used in [Widget constructors](https://github.com/Aylur/ags/blob/bbbf87a21705b7566c2361bc1b978517530d10c0/src/widgets/widget.ts#L249). the above mentioned three properties `emitter`, `prop`, `transformFn` will be "consumed" and be translated into [Widget.bind](https://github.com/Aylur/ags/blob/bbbf87a21705b7566c2361bc1b978517530d10c0/src/widgets/widget.ts#L181) calls which are just again translated into [Widget.hook](https://github.com/Aylur/ags/blob/bbbf87a21705b7566c2361bc1b978517530d10c0/src/widgets/widget.ts#L142) calls

all of this mess to have typesafety through using generics, so we can do
```js
Widget({
  prop: object.bind('prop').transform(prop => prop)
})
```
instead of the basic imperative gobject way
```js
function transform(prop) {
  return `transformed: ${prop}`
}

const widget = Widget({
  prop: transform(object.prop)
})

const id = object.connect('notify::prop', () => {
  widget.prop = transform(object.prop)
})

widget.connect('destroy', () => {
  object.disconnect(id)
})
```
